### PR TITLE
Skip address resolution in a case of exception for CPMemberInfo [HZ-1943]

### PR DESCRIPTION
**Pull request by arodionov**
If members are using internal hostnames that are not known to the client, then deserialization of `Collection<CPMember>` can fail on the client side since the hostname is unknown. 
In that case, we can skip hostname resolution and create the unresolved `Address`.

Fix https://github.com/hazelcast/hazelcast/issues/23311

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/cp/internal/CPMemberInfo.java [hunk 1]
```diff
@@ -86,7 +87,11 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
         endpoint = new RaftEndpointImpl(uuid);
         String host = in.readUTF();
         int port = in.readInt();
-        address = new Address(host, port);
+        try {
+            address = new Address(host, port);
+        } catch (UnknownHostException ex) {
+            address = Address.createUnresolvedAddress(host, port);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/cp/internal/CPMemberInfo.java:93]: can you add a test for serialization/deserialization of `CPMemberInfo`?

Comment (arodionov) [reply hazelcast/src/main/java/com/hazelcast/cp/internal/CPMemberInfo.java:93]: Yes, I will do that.

**Review by ahmetmircik — APPROVED** [review]

File: hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    private static final InternalSerializationService SERIALIZATION_SERVICE =
+            new DefaultSerializationServiceBuilder().build();
+
+    @Test
+    public void testDeserialization() throws UnknownHostException {
+        InetAddress localHost = InetAddress.getLocalHost();
+        Address address = new Address(localHost.getHostName(), 0);
+        CPMemberInfo cpMemberInfo = new CPMemberInfo(UUID.randomUUID(), address);
+
+        // need a list of CPMemberInfo to trigger the default java serializer
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberInfoTest.java:53]: I always thought `Arrays.asList` was returning an ArrayList and I was curious why we are using the Java serializer here, given that we have a default serializer for ArrayList (ArrayListStreamSerializer).

It turns out `Arrays.asList` returns a list named ArrayList, but it is not the ArrayList we have the serializer registered, but another one, that just has the same name. So, we are fine here 👍

File: hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberInfoTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    public void testDeserializationWithUnresolvableHost() {
+        Address unresolvedAddress = Address.createUnresolvedAddress("unresolved-host", 0);
+        CPMemberInfo cpMemberInfo = new CPMemberInfo(UUID.randomUUID(), unresolvedAddress);
+
+        // need a list of CPMemberInfo to trigger the default java serializer
+        Data data = SERIALIZATION_SERVICE.toData(Arrays.asList(cpMemberInfo));
+        List<CPMemberInfo> cpMemberInfoList =  SERIALIZATION_SERVICE.toObject(data);
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/cp/internal/CPMemberInfoTest.java:70]:
```suggestion
        List<CPMemberInfo> cpMemberInfoList = SERIALIZATION_SERVICE.toObject(data);
```

**Review by mdumandag — APPROVED** [review]

**Review by vbekiaris — APPROVED** [review]

**Pull request merged**